### PR TITLE
Add cache utility methods

### PR DESCRIPTION
**Pull request by beatriz**
This introduces a new `MemoizeFn` to simplify the creation of utility methods for the usage of `scalacache` memoization.

The utility's methods accept a method/function to cache/memoize and a configuration object, which guides the instantiation of the underlying cache mechanism. They return a memoized method following the signature of the original method.

### CODE REVIEW

**Review by thyandrecardoso — CHANGES_REQUESTED** [review]
I ran the `undeclaredCompileDependencies` command from the `sbt-explicit-dependencies` sbt plugin and it seems we have some undeclared dependencies on the `caching` project.

Also, should we add some examples from these new features on README?

File: docs/README.md [hunk 1]
```diff
@@ -376,6 +377,47 @@ The `S3Bucket` class wraps an instance of `AmazonS3Client` (from AWS SDK for Jav
...
+## Caching
+The `apso-caching` module provides provides utilities for caching. 
+
+To use it in an existing SBT project, add the following dependency to your `build.sbt`:
+
+```scala
+libraryDependencies += "com.velocidi" %% "apso-caching" % "@VERSION@"
+```
+
+Apso provides utilities to simplify the caching of method calls, with [ScalaCache](https://cb372.github.io/scalacache/)'s `Guava` or `Caffeine`. These are provided in an implicit scope, to make its invocation simpler:
```
Comment (thyandrecardoso) [inline docs/README.md:389]: I would change that a little:
```suggestion
Apso provides utilities to simplify the caching of method calls, with [ScalaCache](https://cb372.github.io/scalacache/) 
and using either `Guava` or `Caffeine` as underlying cache implementations.

These utilities are provided as `.cached()` and `.cachedFn()` extension methods over all `FunctionN[]` types:
```

File: docs/README.md [hunk 2]
```diff
@@ -376,6 +377,47 @@ The `S3Bucket` class wraps an instance of `AmazonS3Client` (from AWS SDK for Jav
...
+
+val x = new AtomicInteger(0)
+
+val cachedFn = ((i: Int) => {
+  val value = x.getAndAdd(i)
+  value
+}).cached(config.Cache.Caffeine(Some(3), None))
+
+cachedFn(2)
+cachedFn(2)
```
Comment (thyandrecardoso) [inline docs/README.md:408]: Nitpick:

```suggestion
cachedFn(1)
cachedFn(1)
x
```

File: docs/README.md [hunk 3]
```diff
@@ -376,6 +377,47 @@ The `S3Bucket` class wraps an instance of `AmazonS3Client` (from AWS SDK for Jav
...
+
+val y = new AtomicInteger(0)
+
+val cachedFutFn = ((i: Int) => Future {
+  val value = y.getAndAdd(i)
+  value
+}).cachedF(config.Cache.Guava(Some(2), None))
+
+Await.result(cachedFutFn(3), Duration.Inf)
+Await.result(cachedFutFn(3), Duration.Inf)
```
Comment (thyandrecardoso) [inline docs/README.md:418]: Nitpick:

```suggestion
Await.result(cachedFutFn(1), Duration.Inf)
Await.result(cachedFutFn(1), Duration.Inf)
y
```

**Review by thyandrecardoso — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 1 above)
Comment (beatriz) [reply docs/README.md]: Applied it on d258cbc8ba4b37f616816c0e7235134778fd07be.

File: docs/README.md (hunk 2 above)
Comment (beatriz) [reply docs/README.md:410]: Applied on d258cbc8ba4b37f616816c0e7235134778fd07be.

File: docs/README.md (hunk 3 above)
Comment (beatriz) [reply docs/README.md:421]: Applied on d258cbc8ba4b37f616816c0e7235134778fd07be.

**Review by thyandrecardoso — APPROVED** [review]

**Pull request merged**